Implement file reading task source

jdm opened the pull request:
Implement the task source API for the File Reader task source, enabling using task sources from non-main threads.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix (partially) #7959 (github issue number if applicable).
- [X] These changes do not require tests because they're refactoring existing code

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12448)

<!-- Reviewable:end -->

KiChjang committed:
Implement file reading task source

And remove superfluous FileReaderEvent enum

jdm committed:
Allow wrapping worker runnables in cancellable runnables.

jdm committed:
Make task queue API usable from non-main threads.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/task_source/file_reading.rs, components/script/dom/bindings/global.rs, components/script/dom/storage.rs, components/script/dom/workerglobalscope.rs, components/script/dom/htmldetailselement.rs, components/script/dom/htmlformelement.rs, components/script/task_source/user_interaction.rs, components/script/dom/filereader.rs, components/script/dom/htmlmediaelement.rs, components/script/dom/htmlimageelement.rs, components/script/script_thread.rs, components/script/dom/window.rs, components/script/task_source/dom_manipulation.rs, components/script/task_source/mod.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

jdm commented:
@KiChjang I rebased your change from #11645 and added supporting code that allows using the file reader task source in workers. Want to review the changes I made?

KiChjang commented on components/script/task_source/mod.rs:
```diff
@@ -8,10 +8,17 @@ pub mod history_traversal;
+                             wrapper: &RunnableWrapper)
+                             -> Result<(), ()>
+                             where T: Runnable + Send + 'static;
+    fn queue<T: Runnable + Send + 'static>(&self, msg: Box<T>, global: GlobalRef) -> Result<(), ()> {
```
This pattern doesn't sit well with me but could be filed as a follow-up issue. I'm not keen on specifying the `GlobalRef` every time a task is queued. Could possibly just use `self.global()`.

KiChjang commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit aa5f34f has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang aa5f34fcd9b83135e46139be50f1e1a199a8b56f -->

bors-servo commented:
:hourglass: Testing commit aa5f34f with merge 48a912f...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1901), [arm32](http://build.servo.org/builders/arm32/builds/2014), [arm64](http://build.servo.org/builders/arm64/builds/2014), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2055), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2175), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2085), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2073), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2144), [windows](http://build.servo.org/builders/windows/builds/1753)
